pkg/daemon/daemon.go

```diff
@@ -900,29 +913,50 @@ func (dn *Daemon) handleNodeUpdate(old, cur interface{}) {
 // prepUpdateFromCluster handles the shared update prepping functionality for
 // flows that expect the cluster to already be available. Returns true if an
 // update is required, false otherwise.
-func (dn *Daemon) prepUpdateFromCluster() (bool, error) {
+func (dn *Daemon) prepUpdateFromCluster() (*mcfgv1.MachineConfig, *mcfgv1.MachineConfig, error) {
 	desiredConfigName, err := getNodeAnnotationExt(dn.node, constants.DesiredMachineConfigAnnotationKey, true)
 	if err != nil {
-		return false, err
+		return nil, nil, err
+	}
+	desiredConfig, err := dn.mcLister.Get(desiredConfigName)
+	if err != nil {
+		return nil, nil, err
 	}
 	// currentConfig is always expected to be there as loadNodeAnnotations
 	// is one of the very first calls when the daemon starts.
 	currentConfigName, err := getNodeAnnotation(dn.node, constants.CurrentMachineConfigAnnotationKey)
 	if err != nil {
-		return false, err
+		return nil, nil, err
+	}
+	currentConfig, err := dn.mcLister.Get(currentConfigName)
+	if err != nil {
+		return nil, nil, err
 	}
 	state, err := getNodeAnnotation(dn.node, constants.MachineConfigDaemonStateAnnotationKey)
 	if err != nil {
-		return false, err
+		return nil, nil, err
+	}
+
+	mcJSON, err := os.Open(currentConfigPath)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer mcJSON.Close()
+	fingerprintMC := &mcfgv1.MachineConfig{}
+	if err := json.NewDecoder(bufio.NewReader(mcJSON)).Decode(fingerprintMC); err != nil {
+		return nil, nil, err
+	}
+	if fingerprintMC.GetName() != currentConfig.GetName() {
+		return fingerprintMC, currentConfig, nil
```

Comment:
So I read your comment a few times...I am finding it really hard to capture the flow/state of things in my head right now.

It feels like this would all be a *lot* simpler to reason about if basically we had `/var/machine-config-daemon/config<hash>.json` and we kept those for every config version that was relevant to us (current, desired) and used it if it was present instead of talking to the api server.

This would also help address the other BZ about having things be deleted from the apiserver.

That said...after thinking about this more I *think* you're right, it will work.  What I was worried about is "if we're saying currentConfig is desiredConfig, how do we then actually honor the real desiredConfig?".  But if they are different then we'll do a config transition and then on the next boot we should transition to the real desiredConfig?

Reply:
OK here's another way to look at it - we're basically saying we can't trust our annotations to describe our current state, the real state is a file on disk.  So why are we looking at the annotation at all, versus setting it to the right thing if it's different from what's on disk?

Reply:
> So why are we looking at the annotation at all, versus setting it to the right thing if it's different from what's on disk?

RIght, I had the same thought indeed, assuming we're talking about the currentConfig (desiredConfig has to come from somewhere/apiserver otherwise we can't really progress right?). So I guess a natural follow up to this would be to avoid talking to the apiserver when querying the currentConfig and just use what it's reflected on disk right?

Reply:
> That said...after thinking about this more I _think_ you're right, it will work. What I was worried about is "if we're saying currentConfig is desiredConfig, how do we then actually honor the real desiredConfig?". But if they are different then we'll do a config transition and then on the next boot we should transition to the real desiredConfig?

right, that's my understand of what will happen if there's a real drift between currentOnAPI vs currentOnDisk.

Basically, the first sync would reconcile currentOnAPI with currentOnDisk, after we're done with that, there will be another immediate sync to honor desiredConfig (which, in our disaster recovery story is gonna be almost always the same as currentOnAPI, unless you backup an etcd with current != desired at the time you actually backup, right?)

Reply:
> It feels like this would all be a _lot_ simpler to reason about if basically we had `/var/machine-config-daemon/config<hash>.json` and we kept those for every config version that was relevant to us (current, desired) and used it if it was present instead of talking to the api server.

I'm not sure how could we keep desired on disk 🤔 that has always have to come from apiserver right? I feel I'm missing something

Reply:
I think my TL;DR though is this code is correct, happy to lgtm it if you want to land as-is; the rework above is optional.

Reply:
> I'm not sure how could we keep desired on disk thinking that has always have to come from apiserver right? I feel I'm missing something

Yeah I edited to ~strikethrough~ that comment.

Reply:
> I think my TL;DR though is this code is correct, happy to lgtm it if you want to land as-is; the rework above is optional.

This commit scares me somehow lol, something I'm going to change cause I Iike it, it's actually using a `current-<hash>.json` naming cause I find it best. Then yeah, I guess we can pull the trigger on this to get things flowing

Reply:
One thing bugging me too is...why don't we

`return fingerprintMC, desiredConfig, nil` ?

Reply:
would make sense since below we return `currentConfig, desiredConfig, nil` in this case, we are using the fingerprintMC as current, we dont need the current I dont think?

Reply:
or are we saying that in _this_ case the currentConfig is the desired from the fingerprint? would it go from fingerprint->current-> desired? or....?

Reply:
> fingerprint->current-> desired? or....?

exactly that but in the DR case, fingerprint=configB,current=configA,desired=configA. So it's just one real sync+reboot

> is there any case where if we have fingerprint, current and desired we would not want to ultimately end up in the desiredConfig?

nope, there shouldn't be any such case afaict. Desired is always evaluated at sync+1 if fingerprint and current drift

Reply:
> One thing bugging me too is...why don't we
> 
> `return fingerprintMC, desiredConfig, nil` ?

Uhm, yeah, so using my flow from above comments, would that work in this case where I take a snapshot exactly in the middle of a sync where current != desired:

```
current=configA != desired=configB
snapshot

upgrade
- current=configB != desired=configC
roll out new configs and MCD syncs
- current=configC == desired=configC

...

trigger a backup restore pre-upgrade which isn't calling informers
- current=configA != desired=configB
but now what's on disk is still configC
```

should  we first go from configC on disk to configA on disk, and only then configB? You code would go straight from configC on disk to configB (which avoids a sync, but should we anyway?)

Reply:
>  (which avoids a sync, but should we anyway or do we risk missing something?)

I am not sure what we'd miss.  It sounds like you're trying to be more conservative here just on general principle?  I don't object to that to be clear.  But I don't understand how this would be different from any other config change.

Reply:
yup, it's really me being just conservative indeed, I guess it would be fine anyway to avoid a sync, so yeah, changing